File: CHAPTER 08. 정적 메서드에서 최상위 함수로/yckim/chapter8/java/v1/Shortlists.java
package chapter8.java.v1;

import chapter8.kotlin.v1.HasPrice;
import chapter8.kotlin.v1.HasRating;
import chapter8.kotlin.v1.HasRelevance;

import java.util.Comparator;
import java.util.List;
import java.util.stream.Stream;

public class Shortlists {

    public static <T> List<T> sorted(
            List<T> shortlist,
            Comparator<? super T> ordering
    ) {
        return shortlist.stream()
                .sorted(ordering)
                .toList();
    }

    public static <T> List<T> removeItemAt(List<T> shortlist, int index) {
        return Stream.concat(
                        shortlist.stream().limit(index),
                        shortlist.stream().skip(index + 1)
                )
                .toList();
    }

    public static Comparator<HasRating> byRating() {
        return Comparator.comparingDouble(HasRating::getRating).reversed();
    }

    public static Comparator<HasPrice> byPriceLowToHigh() {
        return Comparator.comparing(HasPrice::getPrice);
    }

    public static <T extends HasPrice & HasRating> Comparator<T> byValue() {
        return Comparator.comparingDouble((T t) -> t.getRating() / t.getPrice()).reversed();
    }

    public static Comparator<HasRelevance> byRelevance() {
        return Comparator.comparingDouble(HasRelevance::getRelevance).reversed();
    }
}

File: CHAPTER 8. 정적 메서드에서 최상위 함수로/yckim/chapter8/kotlin/v1/Shortlists.kt
package chapter8.kotlin.v1

import java.util.stream.Stream

object Shortlists {
    fun <T> sorted(
        shortlist: List<T>,
        ordering: Comparator<in T>?
    ): List<T> {
        return shortlist.stream()
            .sorted(ordering)
            .toList()
    }

    fun <T> removeItemAt(shortlist: List<T>, index: Int): List<T> {
        return Stream.concat(
            shortlist.stream().limit(index.toLong()),
            shortlist.stream().skip((index + 1).toLong())
        )
            .toList()
    }

    fun byRating(): Comparator<HasRating> {
        return Comparator.comparingDouble(HasRating::rating).reversed()
    }

    fun byPriceLowToHigh(): Comparator<HasPrice> {
        return Comparator.comparing(HasPrice::price)
    }

    fun <T> byValue(): Comparator<T> where T : HasPrice?, T : HasRating? {
        return Comparator.comparingDouble { t: T -> t!!.rating / t.price }.reversed()
    }

    fun byRelevance(): Comparator<HasRelevance> {
        return Comparator.comparingDouble(HasRelevance::relevance).reversed()
    }
}